### ISSUE TITLE
AS3's nested installation task for service-discovery hangs on service restart

### ISSUE THREAD

**Issue by jgruber** [F5Networks/f5-appsvcs-extension#108]
##### ISSUE TYPE
 - Bug Report

##### AS3 BUILD/ VERSION

```
f5-appsvcs-3.12.0-5
```

##### BIGIP VERSION

```
# tmsh show sys version
Sys::Version
Main Package
  Product     BIG-IP
  Version     14.1.0.5
  Build       0.0.5
  Edition     Point Release 5
  Date        Tue May 14 11:15:38 PDT 2019
```

##### SUMMARY

```
Initial GET requests to the /declare endpoint returns errors. restnoded log shows:

Thu, 20 Jun 2019 01:59:00 GMT - severe: [appsvcs] {"message":"Error: Command failed: ls /var/config/rest/iapps/f5-appsvcs/encryptedDeclaration\nls: cannot access /var/config/rest/iapps/f5-appsvcs/encryptedDeclaration: No such file or directory\n","code":"2","level":"error"}
```

##### STEPS TO REPRODUCE

This appears to be a timing issue when a BIG-IP first boots. Difficult to reproduce, and the issue clears after a few minutes and restarting restnoded.

```
No declaration required.. Happen when GET /declare with no existing initial declaration
```

##### EXPECTED RESULTS

A blank declaration (204 No Content)
```
```

##### ACTUAL RESULTS

You get a 404 with this message

```
{
    "code": 404,
    "message": "Public URI path not registered: /shared/appsvcs/declare",
    "referer": "192.168.0.65",
    "restOperationId": 5082395,
    "kind": ":resterrorresponse"
}
```
AS3 is indeed installed and the error clears if you restart restnoded.

**Comment by dstokesf5** [comment]
That log message about not finding an encrypted file should probably be a warning. AS3 should continue just fine without that file, and I have not seen a case where it hasn't. I suspect it is a red herring in this case.

Are you sure AS3 has had enough time to finish starting up? After AS3 is installed, it might take some time before it is ready if it needs to install the service discovery worker. You should see a message in the logs when AS3 is ready:
```
Thu, 13 Jun 2019 15:42:50 GMT - config: [RestWorker] /shared/appsvcs has started. Name:RestWorker
```

**Comment by jgruber** [comment]
When this error happens, if you query the extensions installed, you see f5-appsvcs, but you do not see f5-service-discovery. When you restart restnoded, f5-service-discovery shows up as installed.

The installation task kicked off by AS3 for the f5-service-discovery is hanging...We must have ONE install task which reports FINISHED when install is finished.. not ONE with a background task. There is not way for an installer to query your 'hidden' task is finished. 

What is happening is mcpd is provisioning a process which restarts services while AS3 is trying to copy the rpm file. The install process hangs... 

Again.. since the installer thought the install FINISHED when it rand the install task for appsrvcs (the only RPM it knows about), there is not way for it to know to 'WAIT' for you process to finish before continuing..

This from the restjavad log:

```
[SEVERE][155][24 Jun 2019 16:59:33 UTC][ChildWrapper] Child runner thread was interrupted. Killing child: b53fa510-433a-45d8-ace7-c9cb75386288
[SEVERE][156][24 Jun 2019 16:59:33 UTC][ChildWrapper] Child runner thread was interrupted. Killing child: b53fa510-433a-45d8-ace7-c9cb75386288
[I][157][24 Jun 2019 16:59:33 UTC][ProcessManager] Child process b53fa510-433a-45d8-ace7-c9cb75386288 has been removed from Process Manager
[I][158][24 Jun 2019 16:59:33 UTC][RestFileReceiver] Cancelled, file:/var/config/rest/downloads/f5-service-discovery-1.1.0-1.noarch.rpm
[SEVERE][159][24 Jun 2019 16:59:33 UTC][8100/shared/file-transfer/uploads FileTransferWorker] Transfer failed for /var/config/rest/downloads/f5-service-discovery-1.1.0-1.noarch.rpm with java.util.concurrent.CancellationException
        at com.f5.rest.common.RestFileReceiver.cancel(RestFileReceiver.java:196)
        at com.f5.rest.workers.FileTransferWorker.cleanUpFileTransferUpload(FileTransferWorker.java:305)
        at com.f5.rest.workers.FileTransferWorker.onShutdown(FileTransferWorker.java:287)
        at com.f5.rest.common.RestServer.shutdownWorker(RestServer.java:1053)
        at com.f5.rest.common.RestServer.shutdown(RestServer.java:1010)
        at com.f5.rest.workers.RestWorkerHost$2.run(RestWorkerHost.java:456)
        at java.lang.Thread.run(Thread.java:748)

[I][160][24 Jun 2019 16:59:33 UTC][FileTransferWorker] Caught unexpected exception while cleaning up temporary files, continuing java.util.concurrent.CancellationException
        at com.f5.rest.common.RestFileReceiver.cancel(RestFileReceiver.java:196)
        at com.f5.rest.workers.FileTransferWorker.cleanUpFileTransferUpload(FileTransferWorker.java:305)
        at com.f5.rest.workers.FileTransferWorker.onShutdown(FileTransferWorker.java:287)
        at com.f5.rest.common.RestServer.shutdownWorker(RestServer.java:1053)
        at com.f5.rest.common.RestServer.shutdown(RestServer.java:1010)
        at com.f5.rest.workers.RestWorkerHost$2.run(RestWorkerHost.java:456)
        at java.lang.Thread.run(Thread.java:748)
```
Then it never recovers.

**Comment by codygreen** [comment]
I'm also running into this issue with 1.6.0 when deployed a large number of standalone BIG-IPs TMOS 13.1.1.4 in Azure. It is not consistent across each deployment, some succeed but most do not. I can provide declaration and logs internally if needed.

**Comment by tkam8** [comment]
I'm also seeing this when running a startup script that installs as3, do, ts using below script in GCP: 
```
restcurl -X POST "shared/iapp/package-management-tasks" -d $DATA
```

After restarting restnoded, I see service-discovery appear in the UI. 

```

"code": 404,
--
491 | "kind": ":resterrorresponse",
492 | "message": "Public URI path not registered: /shared/appsvcs/declare",

```

```
# cat /var/log/restnoded/restnoded.log | grep appsvcs
Sat, 04 Jan 2020 11:53:10 GMT - finest: [LoaderWorker] triggered at path:  /var/config/rest/iapps/f5-appsvcs/nodejs/restWorker.js
Sat, 04 Jan 2020 11:53:12 GMT - info: starting /shared/appsvcs
Sat, 04 Jan 2020 11:54:37 GMT - severe: [appsvcs] {"message":"Failed to complete startup. Error: socket hang up\n    at createHangUpError (_http_client.js:331:15)\n    at Socket.socketOnEnd (_http_client.js:423:23)\n    at emitNone (events.js:111:20)\n    at Socket.emit (events.js:208:7)\n    at endReadableNT (_stream_readable.js:1064:12)\n    at _combinedTickCallback (internal/process/next_tick.js:138:11)\n    at process._tickCallback (internal/process/next_tick.js:180:9)","level":"error"}
Sat, 04 Jan 2020 11:54:37 GMT - severe: [RestWorker] [RestWorker] worker failed to load state: [object Object]:shared/appsvcs
```

```
# tmsh show sys software

--------------------------------------------------
Sys::Software Status
Volume  Product   Version  Build  Active    Status
--------------------------------------------------
HD1.1    BIG-IP  14.1.0.1  0.0.7     yes  complete
```

```
# restcurl -u admin: /mgmt/shared/appsvcs/info
{
  "version": "3.16.0",
  "release": "6",
  "schemaCurrent": "3.16.0",
  "schemaMinimum": "3.0.0"
}
```

**Comment by dstokesf5** [comment]
There are two issues in this thread.

The first is that AS3 has to do additional installation tasks after the RPM is installed. The framework gives us no way to do these additional tasks during installation, so AS3 is forced to do them during start up. AS3 has no way to control when the package management API reports AS3 as being successfully installed. The most reliable way for a script to determine that AS3 is ready to go, is by sending GET requests to the /info endpoint.

The second problem, is that the service discover worker installation can fail when a previous upload of the RPM was interrupted. This is being tracked in our product backlog as AUTOTOOL-1078.

**Comment by dstokesf5** [comment]
AUTOTOOL-1078 was resolved 3.17.0. Closing this issue.
